apps/src/pythonlab/patches/pythonScriptUtils.ts

```diff
@@ -23,3 +14,98 @@ export function applyPatches(originalCode: string) {
   }
   return finalCode;
 }
+
+// Pyodide uses the same interpreter for the lifetime of the browser tab.
+// In order to ensure we get updated user code for each run, we delete the
+// modules created by the user code from the sys.modules cache.
+export function deleteCachedUserModules(
+  source: MultiFileSource,
+  excludedFileName: string
+) {
+  let result = '\n';
+  for (const file of Object.values(source.files)) {
+    if (file.name !== excludedFileName) {
+      const filePath = getModuleName(file.id, source);
+      result += `
+if "${filePath}" in sys.modules:
+  del sys.modules['${filePath}']
+`;
+    }
+  }
+  return result;
```

Comment:
This is fine, but as a non-blocking suggestion I hate doing string concatenation like this, especially where the newlines are significant. I'd skew towards building an array and joining it.

```
const result = []
...
if (file.name !== excludedFileName) {
  const filePath = getModuleName(file.id, source);
  result.push(
    [`if "${filePath}" in sys.modules:`, `  del sys.modules['${filePath}']`].join("\n")
  );
}

return result.join("\n")
```

Again, this one's personal preference and not required. I also waffle on whether the embedded python should be joined or kept as a formatted template literal. But the outer list of results feels like it should be in a joined list.

Reply:
I decided to keep the embedded python as a formatted string because it makes it a little easier to read and tell that there's an if statement going on (and because python is picky about tabs). But added the outer array!